File: state/ts/shop-sale/order.ts
import { OrderState } from './order-state.interface';
import { NewState } from './concrete-states';

/**
 * 訂單 class
 */
export class Order {
  private state: OrderState;

  constructor() {
    // Default state is New
    this.state = new NewState(this);
  }

  public setState(state: OrderState): void {
    this.state = state;
  }

  public getState(): OrderState {
    return this.state;
  }

  public payOrder(): void {
    this.state.payOrder();
  }

  public shipOrder(): void {
    this.state.shipOrder();
  }

  public deliverOrder(): void {
    this.state.deliverOrder();
  }

  public cancelOrder(): void {
    this.state.cancelOrder();
  }
}

File: state/java/shopsale/Order.java
package state.java.shopsale;

public class Order {
  private OrderState state;

  public Order() {
    // Default state is New
    this.state = new NewState(this);
  }

  public void setState(OrderState state) {
    this.state = state;
  }

  public OrderState getState() {
    return this.state;
  }

  public void payOrder() {
    this.state.payOrder();
  }

  public void shipOrder() {
    this.state.shipOrder();
  }

  public void deliverOrder() {
    this.state.deliverOrder();
  }

  public void cancelOrder() {
    this.state.cancelOrder();
  }
}
